src/components/AuthForm.tsx

```diff
@@ -0,0 +1,269 @@
+import { app } from "../firebase";
+import { collection, addDoc } from "firebase/firestore";
+import { db } from "../firebase";
+import { object, string } from "yup";
+import { useAuth } from "../contexts/AuthContext";
+import classes from "../styles/AuthForm.module.scss";
+
+interface AuthFormProps {
+  isSignIn: boolean;
+  setIsSignIn: (value: boolean) => void;
+}
+
+interface AuthData {
+  username?: string;
+  email: string;
+  password: string;
+  confirmPassword?: string;
+}
+
+type FormError = {
+  fieldName: keyof AuthData;
+  message: string;
+};
+
+const AuthForm: React.FC<AuthFormProps> = ({ isSignIn, setIsSignIn }) => {
+  const [authData, setAuthData] = useState<AuthData>({
+    email: "",
+    password: "",
+  });
+
+  const [errors, setErrors] = useState<FormError[]>([]);
+  const [error, setError] = useState<string>("");
+
+  const { login } = useAuth();
+
+  useEffect(() => {
+    setAuthData({
+      email: "",
+      password: "",
+      username: "",
+      confirmPassword: "",
+    });
+    setErrors([]);
+    setError("");
+  }, [isSignIn]);
+
+  const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    const { name, value } = e.target;
+    setAuthData({ ...authData, [name]: value });
+  };
+
+  const handleBlur = (e: React.FocusEvent<HTMLInputElement>) => {
+    const { name, value } = e.target;
+    validateField(name, value);
+  };
+
+  const validateField = (name: string, value: string) => {
+    try {
+      const fieldSchema = object().shape({
+        username:
+          !isSignIn && name === "username"
+            ? string().required("Username is required")
+            : string(),
+        email:
+          name === "email"
+            ? string().email("Invalid email").required("Email is required")
+            : string(),
+        password:
+          name === "password"
+            ? string()
+                .min(6, "Password must be at least 6 characters")
+                .required("Password is required")
+            : string(),
+        confirmPassword:
+          !isSignIn && name === "confirmPassword"
+            ? string()
+                .oneOf([authData.password], "Passwords must match")
+                .required("Confirm password is required")
+            : string(),
+      });
+
+      fieldSchema.validateSyncAt(name, { [name]: value });
+      setErrors((errors) => errors.filter((error) => error.fieldName !== name));
+    } catch (error: any) {
+      setErrors((errors) => [
+        ...errors.filter((err) => err.fieldName !== name),
+        { fieldName: name as keyof AuthData, message: error.message },
+      ]);
+    }
+  };
+
+  const handleSubmit = async (e: React.FormEvent) => {
+    e.preventDefault();
+    try {
+      const schema = object().shape({
+        username: !isSignIn
+          ? string().required("Username is required")
+          : string().notRequired(),
+        email: string().email("Invalid email").required("Email is required"),
```

Comment:
I understand that you are trying to validate if the user tries to submit an empty form, but I would just disable submit button if there are no inputs and no errors. lets discuss this